File: src/FlashcardParser/CategoriesFinder.ts
import {
    CATEGORIES,
    COMMENT_SYMBOL,
    CATEGORY_START_SYMBOL,
    CATEGORY_END_SYMBOL
} from './Symbols';

export interface Category {
    name: string;
    line: number;
}

/**
 * Reads a list of lines and looks for every category.
 * @param lines 
 * @returns A sorted list of (category, lineNumber) elements.
 */
export function categoriesOf(lines: string[]): Array<Category> {
    const categories: Array<Category> = [];
    const knownCategories: ReadonlyArray<string> = Object.values(CATEGORIES) as ReadonlyArray<string>;

    try {
        for (let line = 0; line < lines.length; line++) {
            const trimmedLine = lines[line].trim();
            if (trimmedLine.startsWith(COMMENT_SYMBOL)) continue;
            if (!trimmedLine.startsWith(CATEGORY_START_SYMBOL)) continue;
            if (!trimmedLine.endsWith(CATEGORY_END_SYMBOL)) continue;

            const name = trimmedLine.substring(1, trimmedLine.length - 1) as string;
            categories.push({ name, line });

            if (!knownCategories.includes(name)) {
                console.warn('[categoriesFinder] Categories not previously known found: ' + name);
            }
        }

    } catch (error) {
        console.error(`[categoriesFinder] ERROR: ${error}`);
    }

    return categories;
}

File: service/flashcard_parser/categories_finder.py
from dataclasses import dataclass
from typing import List

from .symbols import (
    COMMENT_SYMBOL,
    CATEGORY_START_SYMBOL,
    CATEGORY_END_SYMBOL,
    CATEGORIES,
)

@dataclass
class Category:
    name: str
    line: int


def categories_of(lines: List[str]) -> List[Category]:
    categories: List[Category] = []
    known_categories = set(CATEGORIES.values())
    for line_num, raw in enumerate(lines):
        trimmed = raw.strip()
        if trimmed.startswith(COMMENT_SYMBOL):
            continue
        if not trimmed.startswith(CATEGORY_START_SYMBOL):
            continue
        if not trimmed.endswith(CATEGORY_END_SYMBOL):
            continue
        name = trimmed[1:-1]
        categories.append(Category(name=name, line=line_num))
        if name not in known_categories:
            print(f"[categoriesFinder] Categories not previously known found: {name}")
    return categories
